skiko/src/nativeJsMain/cpp/shaper/Shaper.cc

```diff
@@ -55,12 +58,92 @@ SKIKO_EXPORT KNativePointer org_jetbrains_skia_shaper_Shaper__1nMake
+    ));
+    if (!fontRunIter) return 0;
+
+    uint8_t defaultBiDiLevel = isLeftToRight ? UBIDI_DEFAULT_LTR : UBIDI_DEFAULT_RTL;
+    std::unique_ptr<SkShaper::BiDiRunIterator> bidiRunIter(SkShaper::MakeBiDiRunIterator(text.c_str(), text.size(), defaultBiDiLevel));
+    if (!bidiRunIter) return 0;
+
+    std::unique_ptr<SkShaper::ScriptRunIterator> scriptRunIter(SkShaper::MakeHbIcuScriptRunIterator(text.c_str(), text.size()));
+    if (!scriptRunIter) return 0;
+
+    std::unique_ptr<SkShaper::LanguageRunIterator> languageRunIter(SkShaper::MakeStdLanguageRunIterator(text.c_str(), text.size()));
+    if (!languageRunIter) return 0;
+
+    SkTextBlobBuilderRunHandler rh(text.c_str(), {offsetX, offsetY});
+    instance->shape(text.c_str(), text.size(), *fontRunIter, *bidiRunIter, *scriptRunIter, *languageRunIter, features.data(), features.size(), width, &rh);
+    SkTextBlob* blob = rh.makeBlob().release();
+
+    return reinterpret_cast<KNativePointer>(blob);
 }
+
 SKIKO_EXPORT KNativePointer org_jetbrains_skia_shaper_Shaper__1nShapeLine
-  (KNativePointer ptr, KInteropPointer textObj, KNativePointer fontPtr, KInteropPointer opts) {
-    TODO("implement org_jetbrains_skia_shaper_Shaper__1nShapeLine");
+  (KNativePointer ptr, KNativePointer textManagedStringPtr, KNativePointer fontPtr, KInt optsFeaturesLen, KInt* optsFeatures, KInt optsBooleanProps) {
+    SkShaper* instance = reinterpret_cast<SkShaper*>(ptr);
+
+    SkString text = *(reinterpret_cast<SkString*>(textManagedStringPtr));
+    std::shared_ptr<UBreakIterator> graphemeIter = skija::shaper::graphemeBreakIterator(text);
+    if (!graphemeIter) return 0;
+
+    SkFont* font = reinterpret_cast<SkFont*>(fontPtr);
+    std::vector<SkShaper::Feature> features = skija::shaper::ShapingOptions::getFeaturesFromIntsArray(optsFeatures, optsFeaturesLen);
+
+    bool aproximatePunctuation = optsBooleanProps & 0x01;
+    bool aproximateSpaces = optsBooleanProps & 0x02;
+    bool isLeftToRight = optsBooleanProps & 0x04;
+
+    std::unique_ptr<SkShaper::FontRunIterator> fontRunIter(new FontRunIterator(
+        text.c_str(),
+        text.size(),
+        *font,
+        SkFontMgr::RefDefault(),
+        graphemeIter,
+        aproximateSpaces,
+        aproximatePunctuation));
+    if (!fontRunIter) return 0;
+
+    uint8_t defaultBiDiLevel = isLeftToRight ? UBIDI_DEFAULT_LTR : UBIDI_DEFAULT_RTL;
+    std::unique_ptr<SkShaper::BiDiRunIterator> bidiRunIter(SkShaper::MakeBiDiRunIterator(text.c_str(), text.size(), defaultBiDiLevel));
+    if (!bidiRunIter) return 0;
+
+    std::unique_ptr<SkShaper::ScriptRunIterator> scriptRunIter(SkShaper::MakeHbIcuScriptRunIterator(text.c_str(), text.size()));
+    if (!scriptRunIter) return 0;
+
+    std::unique_ptr<SkShaper::LanguageRunIterator> languageRunIter(SkShaper::MakeStdLanguageRunIterator(text.c_str(), text.size()));
+    if (!languageRunIter) return 0;
+
+    TextLine* line;
+    if (text.size() == 0)
+        line = new TextLine(*font);
```

Comment:
Can we return it earlier to avoid creation of unnecessary objects

Reply:
fixed